airsonic-main/src/main/java/org/airsonic/player/service/search/IndexManager.java

```diff
@@ -1,169 +1,236 @@
+            } else {
+                Document document = documentFactory.createArtistDocument(mediaFile);
+                writers.get(IndexType.ARTIST).updateDocument(primarykey, document);
+            }
+        } catch (Exception x) {
+            LOG.error("Failed to create search index for " + mediaFile, x);
+        }
+    }
+
+    private String getIndexRootDirectoryName() {
+        return INDEX_DIR;
+    }
+
+    public final void startIndexing() {
+        try {
+            for (IndexType IndexType : IndexType.values()) {
+                writers.put(IndexType, createIndexWriter(IndexType));
+            }
+        } catch (IOException e) {
+            LOG.error("Failed to create search index.", e);
+        }
+    }
+
+    /**
+     * 
+     * @param indexType
+     * @return
+     * @throws IOException
+     */
+    private IndexWriter createIndexWriter(IndexType indexType) throws IOException {
+        File dir = util.getDirectory.apply(getIndexRootDirectoryName(), indexType);
+        IndexWriterConfig config = new IndexWriterConfig(analyzerFactory.getAnalyzer());
+        return new IndexWriter(FSDirectory.open(dir.toPath()), config);
+    }
+
+    /**
+     * Close Writer of all indexes and update SearcherManager.
+     * Called at the end of the Scan flow.
+     */
+    public void stopIndexing() {
+        Arrays.asList(IndexType.values()).forEach(i -> stopIndexing(i));
+    }
+
+    /**
+     * Close Writer of specified index and refresh SearcherManager.
+     * @param type
+     */
+    private void stopIndexing(IndexType type) {
+
+        boolean isUpdate = false;
+        // close
+        try {
+            isUpdate = -1 == writers.get(type).commit();
+            writers.get(type).close();
+            writers.remove(type);
+            LOG.trace("Success to create or update search index : [" + type + "]");
+        } catch (IOException e) {
+            LOG.error("Failed to create search index.", e);
+        } finally {
+            FileUtil.closeQuietly(writers.get(type));
+        }
+
+        // refresh reader as index may have been written
+        if (isUpdate && searchers.containsKey(type)) {
+            try {
+                searchers.get(type).maybeRefresh();
+                LOG.trace("SearcherManager has been refreshed : [" + type + "]");
+            } catch (IOException e) {
+                LOG.error("Failed to refresh SearcherManager : [" + type + "]", e);
+                searchers.remove(type);
+            }
+        }
+
+    }
+
+    /**
+     * Return the IndexSearcher of the specified index.
+     * At initial startup, it may return null
+     * if the user performs any search before performing a scan.
+     * 
+     * @param indexType
+     * @return
+     */
+    public @Nullable IndexSearcher getSearcher(IndexType indexType) {
+        if (!searchers.containsKey(indexType)) {
+            synchronized (searchers) {
```

Comment:
Instead of using synchronized directly, can you instead use `computeIfAbsent`. Then you don't have to worry about putting the synchronized in the right spot (I don't think it is here).

Reply:
It's exactly as you said.
Furthermore, getSearcher is beautiful for Optional.


The barrier is lambda's exception handling.
We can do more things by writing lambdas that throw exceptions.
However, in the vicinity of search there are a lot of IOExceptions.
And if we introduce it, the source code will be quite different.

After thinking for a while, I prioritized the old try catch and avoid adopting new mechanisms sharply.
If a common view of Airsonic is shown for lambda exceptions, I rewrite this PR and follow it in all subsequent code.
This feels like a policy that is included in everything else, not just this PR.

I've removed synchronized.